File: typescript/examples/07_medical_ai.ts
/**
 * Parmana TypeScript SDK
 *
 * Example 07
 *
 * Medical AI
 *
 * Demonstrates how Parmana governs AI-assisted
 * clinical decision execution while preserving
 * an immutable execution trust chain.
 */

import {
  Authority,
  Authorization,
  BusinessTransaction,
  DecisionOutcome,
  Execution,
  ExecutionMode,
  ExecutionStatus,
  ExecutionTrustRecord,
  Intent,
  ParmanaClient,
  PolicyReference,
} from "../src/index.js";

async function main(): Promise<void> {
  const client = new ParmanaClient({
    endpoint: "http://localhost:8080",
  });

  const authority: Authority = {
    authorityId: "hospital-001",
    authorityName: "City General Hospital",
    createdAt: new Date(),
  };

  const authorization: Authorization = {
    authorizationId: "authorization-001",
    authorityId: authority.authorityId,
    subject: "clinical-ai-assistant",
    permissions: [
      "CLINICAL_DECISION_SUPPORT",
    ],
    issuedAt: new Date(),
    expiresAt: new Date(
      Date.now() + 365 * 24 * 60 * 60 * 1000,
    ),
  };

  const intent: Intent = {
    intentId: "intent-001",
    authorizationId: authorization.authorizationId,
    operation: "ASSESS_PATIENT",
    target: "patient-100234",
    createdAt: new Date(),
  };

  const policy: PolicyReference = {
    policyName: "clinical-decision-policy",
    policyVersion: "1.0.0",
  };

  const transaction: BusinessTransaction = {
    businessTransactionId: "medical-txn-001",
    authority,
    authorization,
    intent,
    policy,
    createdAt: new Date(),
  };

  const execution: Execution = {
    executionId: "execution-001",

    businessTransactionId:
      transaction.businessTransactionId,

    decision: {
      decisionId: "decision-001",

      intentId: intent.intentId,

      policy,

      signals: {
        patientAge: 64,
        bloodPressure: "145/90",
        heartRate: 96,
        oxygenSaturation: 98,
        allergyCheck: "CLEAR",
        physicianAvailable: true,
      },

      outcome: DecisionOutcome.APPROVED,

      reason:
        "Clinical policy satisfied. AI recommendation may be presented to physician.",

      evaluatedAt: new Date(),
    },

    status: ExecutionStatus.COMPLETED,

    mode: ExecutionMode.SYNC,

    startedAt: new Date(),

    completedAt: new Date(),

    evidence: {
      diagnosisSuggestion:
        "Community Acquired Pneumonia",

      confidence: 0.94,

      physicianReviewRequired: true,

      physicianApproved: true,
    },
  };

  const trustRecord: ExecutionTrustRecord = {
    trustRecordId: "trust-record-medical-001",

    businessTransactionId:
      transaction.businessTransactionId,

    transaction,

    overrides: [],

    executions: [
      execution,
    ],

    verifications: [],

    receipts: [],

    trustRecordHash:
      "d71f2dcb4d3044bb94752c13abfe8129",

    createdAt: new Date(),

    updatedAt: new Date(),
  };

  console.log();
  console.log("======================================");
  console.log(" Parmana TypeScript SDK");
  console.log(" Example 07 - Medical AI");
  console.log("======================================");
  console.log();

  console.log("Clinical Case");
  console.log("------------------------------");
  console.log(
    "Hospital     :",
    authority.authorityName,
  );
  console.log(
    "Patient      :",
    intent.target,
  );
  console.log(
    "Operation    :",
    intent.operation,
  );

  console.log();

  console.log("Policy");
  console.log("------------------------------");
  console.log(
    "Policy       :",
    policy.policyName,
  );
  console.log(
    "Version      :",
    policy.policyVersion,
  );

  console.log();

  console.log("Decision");
  console.log("------------------------------");
  console.log(
    "Outcome      :",
    execution.decision.outcome,
  );
  console.log(
    "Reason       :",
    execution.decision.reason,
  );

  console.log();

  console.log("Clinical Signals");
  console.log("------------------------------");
  console.log(
    execution.decision.signals,
  );

  console.log();

  console.log("Execution Evidence");
  console.log("------------------------------");
  console.log(
    execution.evidence,
  );

  console.log();

  console.log("Execution Trust Record");
  console.log("------------------------------");
  console.log(
    "Trust Record :",
    trustRecord.trustRecordId,
  );
  console.log(
    "Hash         :",
    trustRecord.trustRecordHash,
  );

  console.log();

  console.log(
    "Submitting transaction to Parmana Runtime...",
  );

  try {
    const receipt = await client.execute(
      transaction,
    );

    console.log();

    console.log("Receipt");
    console.log("------------------------------");
    console.log(
      "Receipt ID   :",
      receipt.receiptId,
    );
    console.log(
      "Algorithm    :",
      receipt.algorithm,
    );
    console.log(
      "Issued At    :",
      receipt.issuedAt,
    );
  } catch (error) {
    console.log();
    console.log(
      "Runtime not available (expected during SDK development).",
    );

    if (error instanceof Error) {
      console.log(error.message);
    }
  }

  console.log();

  console.log("Execution Trust Chain");
  console.log("------------------------------");
  console.log("✓ Hospital Authority");
  console.log("✓ Clinical Authorization");
  console.log("✓ Medical Intent");
  console.log("✓ Policy Reference");
  console.log("✓ Clinical Signals");
  console.log("✓ Policy Decision");
  console.log("✓ AI Recommendation");
  console.log("✓ Execution Evidence");
  console.log("✓ Execution Trust Record");

  console.log();

  console.log(
    "Parmana does not diagnose patients.",
  );
  console.log(
    "It governs the execution of AI-assisted",
  );
  console.log(
    "clinical workflows so every authorization,",
  );
  console.log(
    "policy evaluation, execution, and outcome",
  );
  console.log(
    "remain independently verifiable.",
  );

  console.log();
  console.log("Example completed.");
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: python-old/examples/07_medical_ai.py
"""
Parmana Python SDK

Example 07

Medical AI

Demonstrates governance of an AI-assisted
clinical recommendation.

Parmana does not diagnose patients.

Parmana ensures the recommendation was:

Authority
        ↓
Authorization
        ↓
Clinical Intent
        ↓
Clinical Policy
        ↓
Decision
        ↓
Physician Approval
        ↓
Execution
        ↓
Verification
        ↓
Receipt
"""

from datetime import datetime
from uuid import uuid4

from parmana import ParmanaClient

from parmana.models.authority import Authority
from parmana.models.authorization import Authorization
from parmana.models.business_transaction import BusinessTransaction
from parmana.models.decision import (
    Decision,
    DecisionOutcome,
)
from parmana.models.execution import (
    Execution,
    ExecutionMode,
    ExecutionStatus,
)
from parmana.models.execution_trust_record import (
    ExecutionTrustRecord,
)
from parmana.models.intent import Intent
from parmana.models.override import Override
from parmana.models.policy_reference import PolicyReference
from parmana.models.receipt import Receipt
from parmana.models.verification import (
    Verification,
    VerificationStatus,
)


def main() -> None:

    #
    # Hospital Authority
    #
    authority = Authority(
        authority_id=str(uuid4()),
        authority_type="HOSPITAL",
        authority_name="City General Hospital",
        created_at=datetime.utcnow(),
        public_key="hospital-public-key",
        signature_algorithm="Ed25519",
    )

    #
    # Physician Authorization
    #
    authorization = Authorization(
        authorization_id=str(uuid4()),
        authority_id=authority.authority_id,
        subject="physician-001",
        action="APPROVE_TREATMENT",
        resource="patient-record",
        issued_at=datetime.utcnow(),
    )

    #
    # AI Recommendation
    #
    intent = Intent(
        intent_id=str(uuid4()),
        operation="RECOMMEND_TREATMENT",
        target="patient-12345",
        parameters={
            "diagnosis": "Community Acquired Pneumonia",
            "recommended_antibiotic": "Amoxicillin",
            "confidence": 0.96,
        },
    )

    #
    # Clinical Policy
    #
    policy = PolicyReference(
        policy_name="clinical-treatment-policy",
        policy_version="2026.1",
    )

    transaction = BusinessTransaction(
        business_transaction_id=str(uuid4()),
        authority=authority,
        authorization=authorization,
        intent=intent,
        policy_reference=policy,
        created_at=datetime.utcnow(),
    )

    #
    # Policy Evaluation
    #
    decision = Decision(
        decision_id=str(uuid4()),
        intent_id=intent.intent_id,
        policy=policy,
        signals={
            "patient_age": 42,
            "drug_allergy": False,
            "renal_function": "NORMAL",
            "pregnant": False,
        },
        outcome=DecisionOutcome.APPROVED,
        reason="Recommendation complies with treatment policy.",
        evaluated_at=datetime.utcnow(),
    )

    #
    # Physician Approval
    #
    override = Override(
        override_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        approved_by="Dr. Sarah Johnson",
        reason="Treatment approved.",
        justification="Clinical review completed.",
        approved_at=datetime.utcnow(),
    )

    #
    # Execution
    #
    execution = Execution(
        execution_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        decision=decision,
        status=ExecutionStatus.COMPLETED,
        mode=ExecutionMode.SYNC,
        started_at=datetime.utcnow(),
        completed_at=datetime.utcnow(),
        evidence={
            "prescription_created": True,
            "ehr_updated": True,
            "physician_signed": True,
        },
    )

    verification = Verification(
        verification_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        status=VerificationStatus.VERIFIED,
        message="Clinical workflow verified.",
        verified_at=datetime.utcnow(),
        trust_record_hash="clinical-trust-record",
    )

    receipt = Receipt(
        receipt_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        execution_id=execution.execution_id,
        trust_record_hash="clinical-trust-record",
        receipt_hash="receipt-hash",
        signature="hospital-signature",
        algorithm="Ed25519",
        issued_at=datetime.utcnow(),
    )

    record = ExecutionTrustRecord(
        trust_record_id=str(uuid4()),
        business_transaction_id=transaction.business_transaction_id,
        transaction=transaction,
        overrides=(override,),
        executions=(execution,),
        verifications=(verification,),
        receipts=(receipt,),
        trust_record_hash="clinical-trust-record",
        created_at=datetime.utcnow(),
        updated_at=datetime.utcnow(),
    )

    client = ParmanaClient()

    verification = client.verify(record)

    print()
    print("==========================================")
    print("Parmana Medical AI")
    print("==========================================")

    print(f"Hospital          : {authority.authority_name}")
    print(f"Physician         : {override.approved_by}")
    print(f"Patient           : {intent.target}")

    print()

    print("AI Recommendation")

    print("-----------------------------")

    print(
        f"Diagnosis         : "
        f"{intent.parameters['diagnosis']}"
    )

    print(
        f"Treatment         : "
        f"{intent.parameters['recommended_antibiotic']}"
    )

    print(
        f"Confidence        : "
        f"{intent.parameters['confidence']}"
    )

    print()

    print("Governance")

    print("-----------------------------")

    print(f"Decision          : {decision.outcome.value}")
    print(f"Policy            : {policy.policy_name}")
    print(f"Physician Review  : Approved")

    print()

    print("Verification")

    print("-----------------------------")

    print(f"Status            : {verification.status.value}")

    print()

    print(
        "Clinical recommendation executed "
        "with a complete execution trust chain."
    )

    print("==========================================")


if __name__ == "__main__":
    main()